### PR TITLE
Pulsar

### PR DESCRIPTION
**Pull request by gunnarmorling**
Based on #27.

@jpechane Could you take a look at this one? For some reason I'm not getting any notifications from Postgres (i.e. seems unrelated to Pulsar). I feel I'm doing something stupid... Any idea what could be wrong here? Thanks!

### CODE REVIEW

File: apache-pulsar/src/main/java/io/debezium/examples/apache/pulsar/config/PropertyLoader.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+public class PropertyLoader {
+    private static Logger logger = Logger.getLogger(PropertyLoader.class);
+
+
+    public static void loadEnvironmentValues(Properties properties) {
+        Enumeration keys = properties.propertyNames();
+        while (keys.hasMoreElements()) {
+            String key = (String) keys.nextElement();
+            String env_safe_key = key.replace(".", "_");
```
Comment (jpechane) [inline apache-pulsar/src/main/java/io/debezium/examples/apache/pulsar/config/PropertyLoader.java]: Please use Java naming confention

Comment (jpechane) [reply apache-pulsar/src/main/java/io/debezium/examples/apache/pulsar/config/PropertyLoader.java]: env_safe_key can use concatenated calls of `replace` like `replace(".", "_").replace("-", "_"). Also char version of method might be better used then String arg version of `replace method

File: apache-pulsar/src/main/java/io/debezium/examples/apache/pulsar/config/PropertyLoader.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+            env_safe_key = env_safe_key.replace("-", "_");
+            String envValue = System.getenv(env_safe_key.toUpperCase());
+            String systemPropValue = System.getProperty(key.toUpperCase());
+            String oldValue = properties.getProperty(key);
+            System.out.println(key + ":" + oldValue);
```
Comment (jpechane) [inline apache-pulsar/src/main/java/io/debezium/examples/apache/pulsar/config/PropertyLoader.java]: Please use logger from the class

File: apache-pulsar/src/main/java/io/debezium/examples/apache/pulsar/config/PropertyLoader.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+            if (envValue != null) {
+                properties.setProperty(key, envValue);
+                logger.debug("Setting java property for key:" + key + " ,value:" + envValue + " ,oldValue:" + oldValue);
```
Comment (jpechane) [inline apache-pulsar/src/main/java/io/debezium/examples/apache/pulsar/config/PropertyLoader.java]: Use slf4j approach to logging

File: apache-pulsar/src/main/java/io/debezium/examples/apache/pulsar/config/PropertyLoader.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+package io.debezium.examples.apache.pulsar.config;
+
+import org.apache.log4j.Logger;
+
+import java.util.Enumeration;
+import java.util.Properties;
+
+
+public class PropertyLoader {
+    private static Logger logger = Logger.getLogger(PropertyLoader.class);
```
Comment (jpechane) [inline apache-pulsar/src/main/java/io/debezium/examples/apache/pulsar/config/PropertyLoader.java]: Do not mix slf4j and plain log4j logging in a single project

File: apache-pulsar/src/main/java/io/debezium/examples/apache/pulsar/PulsarProducer.java [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+        awaitTermination(executor);
+    }
+
+    /**
+     * For Every record this method will be invoked* @param record
+     */
+    private void sendRecord(SourceRecord record) {
+        final byte[] payload = valueConverter.fromConnectData("dummy", record.valueSchema(), record.value());
+        final byte[] key = keyConverter.fromConnectData("dummy", record.keySchema(), record.key());
+        logger.debug("Publishing Topic --> " + record.topic());
```
Comment (jpechane) [inline apache-pulsar/src/main/java/io/debezium/examples/apache/pulsar/PulsarProducer.java]: Please use log4j logging approach

File: apache-pulsar/src/main/java/io/debezium/examples/apache/pulsar/PulsarProducer.java [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+            catch (PulsarClientException e) {
+                throw new RuntimeException(e);
+            }
+        }));
+
+        awaitTermination(executor);
+    }
+
+    /**
+     * For Every record this method will be invoked* @param record
```
Comment (jpechane) [inline apache-pulsar/src/main/java/io/debezium/examples/apache/pulsar/PulsarProducer.java]: Capital `E`, `@param` should go on its own line

File: apache-pulsar/src/main/java/io/debezium/examples/apache/pulsar/PulsarProducer.java [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+
+public class PulsarProducer implements Runnable {
+
+    private static final Logger logger = LoggerFactory.getLogger(PulsarProducer.class);
+
+    private final Configuration config;
+    private final JsonConverter valueConverter;
+    private final JsonConverter keyConverter;
+    private PulsarClient client;
+    private final HashMap<String, Producer<String>> producerHashMap = new HashMap<>();
```
Comment (jpechane) [inline apache-pulsar/src/main/java/io/debezium/examples/apache/pulsar/PulsarProducer.java]: Neitther type nor name should contain the implementation of the map, so the hash should be dropped

File: apache-pulsar/src/main/java/io/debezium/examples/apache/pulsar/PulsarProducer.java [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+        keyConverter = new JsonConverter();
+        keyConverter.configure(config.asMap(), true);
+        valueConverter = new JsonConverter();
+        valueConverter.configure(config.asMap(), false);
+    }
+
+    private Producer<String> getProducer(String topic) {
+        String topicFormat = propConfig.getProperty("pulsar.topic");
+        String topicURI = MessageFormat.format(topicFormat, topic);
+        Producer<String> producer = producerHashMap.get(topic);
```
Comment (jpechane) [inline apache-pulsar/src/main/java/io/debezium/examples/apache/pulsar/PulsarProducer.java]: Use `computeIfAbsent` to simplify the code

File: apache-pulsar/src/main/java/io/debezium/examples/apache/pulsar/PulsarProducer.java [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+        final byte[] payload = valueConverter.fromConnectData("dummy", record.valueSchema(), record.value());
+        final byte[] key = keyConverter.fromConnectData("dummy", record.keySchema(), record.key());
+        logger.debug("Publishing Topic --> " + record.topic());
+        logger.debug("Key -->" + new String(key));
+        logger.debug("Payload --> " + new String(payload));
+        Producer<String> producer;
+
+        producer = getProducer(record.topic());
+
+        while (true) {
```
Comment (jpechane) [inline apache-pulsar/src/main/java/io/debezium/examples/apache/pulsar/PulsarProducer.java]: I do not think the `while` loop is necessary here

**Comment by gunnarmorling** [comment]
@jpechane Thanks for reviewing. FWIW, I wasn't too hung up on the style issues, as it's "just" demo. More curious why I don't receive any change events from PG ;)

File: apache-pulsar/README.md [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+    docker run -it --rm --name postgres -p 5432:5432 -e POSTGRES_USER=postgres -e POSTGRES_PASSWORD=postgres debezium/example-postgres:0.8
+
+* Launch psql to run some SQL queries:
+
+    docker run -it --rm -e PGOPTIONS="--search_path=inventory" -e PGPASSWORD=postgres --link postgres:postgres debezium/example-postgres:0.8 psql -h postgres -U postgres
+
+## Building the Source Code and Running Debezium Embedded
+
+    mvn clean install
+    docker exec -i pulsar bin/pulsar-client consume -s mysubscription persistent://public/default/test.inventory.customers
```
Comment (jpechane) [inline apache-pulsar/README.md]: Add `-n 0` option to consume message in loop

Comment (jpechane) [inline apache-pulsar/README.md]: Fix command that starts Debezium, the correct one is probably `java -jar target/apache-pulsar-1.0-SNAPSHOT-jar-with-dependencies.jar `

File: apache-pulsar/README.md [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+## Testing
+
+Modify a record in psql, e.g. like this:
+
+    update customers set first_name = 'Sarah' where id = 1001;
+
+Consume the corresponding change event from the Pulsar topic, e.g. using the command line client
+(by default it will consume one event and then return):
+
+    docker exec -i pulsar bin/pulsar-client consume -s my-subscription persistent://public/default/test.inventory.customers
```
Comment (jpechane) [inline apache-pulsar/README.md]: Add `-n 0` option to consume message in loop

**Comment by jpechane** [comment]
@gunnarmorling It works for me, have you started the DBZ-> Pulsar app? That step was missing in the docs. Instead it was starting the client reading from the Pulsar instance.
I've also add few comments for improvement.

File: apache-pulsar/README.md (hunk 10 above)
Comment (gunnarmorling) [reply apache-pulsar/README.md]: Ah, yes. Thx!

File: apache-pulsar/README.md (hunk 11 above)
Comment (gunnarmorling) [reply apache-pulsar/README.md]: Thx, great improvement!

File: apache-pulsar/src/main/java/io/debezium/examples/apache/pulsar/PulsarProducer.java (hunk 9 above)
Comment (gunnarmorling) [reply apache-pulsar/src/main/java/io/debezium/examples/apache/pulsar/PulsarProducer.java]: I think that's fine.

File: apache-pulsar/src/main/java/io/debezium/examples/apache/pulsar/PulsarProducer.java (hunk 7 above)
Comment (gunnarmorling) [reply apache-pulsar/src/main/java/io/debezium/examples/apache/pulsar/PulsarProducer.java]: Ok :)

File: apache-pulsar/src/main/java/io/debezium/examples/apache/pulsar/PulsarProducer.java (hunk 8 above)
Comment (gunnarmorling) [reply apache-pulsar/src/main/java/io/debezium/examples/apache/pulsar/PulsarProducer.java]: Done.

File: apache-pulsar/src/main/java/io/debezium/examples/apache/pulsar/PulsarProducer.java (hunk 6 above)
Comment (gunnarmorling) [reply apache-pulsar/src/main/java/io/debezium/examples/apache/pulsar/PulsarProducer.java]: Fixed.

File: apache-pulsar/src/main/java/io/debezium/examples/apache/pulsar/PulsarProducer.java (hunk 5 above)
Comment (gunnarmorling) [reply apache-pulsar/src/main/java/io/debezium/examples/apache/pulsar/PulsarProducer.java]: Done.

File: apache-pulsar/src/main/resources/log4j.properties [hunk 12]
```diff
@@ -0,0 +1,10 @@
+# Root logger option
+log4j.rootLogger=INFO, stdout
+
+# Redirect log messages to console
+log4j.appender.stdout=org.apache.log4j.ConsoleAppender
+log4j.appender.stdout.Target=System.out
+log4j.appender.stdout.layout=org.apache.log4j.PatternLayout
+log4j.appender.stdout.layout.ConversionPattern=[%d{yyyy-MM-dd}T%d{HH:mm:ss.SSS}Z] %p %t %c{1}:%L %m%n
+log4j.logger.in.hack4geek.connector.dbz.pulsar=DEBUG
```
[the remark this answers is not in the document]
Comment (gunnarmorling) [reply apache-pulsar/src/main/resources/log4j.properties]: Good catch, fixed.

File: apache-pulsar/src/main/java/io/debezium/examples/apache/pulsar/config/PropertyLoader.java [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+            String env_safe_key = key.replace(".", "_");
+            env_safe_key = env_safe_key.replace("-", "_");
+            String envValue = System.getenv(env_safe_key.toUpperCase());
+            String systemPropValue = System.getProperty(key.toUpperCase());
+            String oldValue = properties.getProperty(key);
+            System.out.println(key + ":" + oldValue);
+            if (envValue != null) {
+                properties.setProperty(key, envValue);
+                logger.debug("Setting java property for key:" + key + " ,value:" + envValue + " ,oldValue:" + oldValue);
+            } else if (systemPropValue != null) {
```
[the remark this answers is not in the document]
Comment (gunnarmorling) [reply apache-pulsar/src/main/java/io/debezium/examples/apache/pulsar/config/PropertyLoader.java]: In this case it's about props loaded from a file, so it makes sense to override them with env variables. That all being said, Debezium's config mechanism already would have supported overriding via system properties, but I'll leave that all as is.

File: apache-pulsar/src/main/java/io/debezium/examples/apache/pulsar/config/PropertyLoader.java (hunk 4 above)
Comment (gunnarmorling) [reply apache-pulsar/src/main/java/io/debezium/examples/apache/pulsar/config/PropertyLoader.java]: Done.

File: apache-pulsar/src/main/java/io/debezium/examples/apache/pulsar/config/PropertyLoader.java (hunk 3 above)
Comment (gunnarmorling) [reply apache-pulsar/src/main/java/io/debezium/examples/apache/pulsar/config/PropertyLoader.java]: Done.

File: apache-pulsar/src/main/java/io/debezium/examples/apache/pulsar/config/PropertyLoader.java (hunk 2 above)
Comment (gunnarmorling) [reply apache-pulsar/src/main/java/io/debezium/examples/apache/pulsar/config/PropertyLoader.java]: Done.

File: apache-pulsar/src/main/java/io/debezium/examples/apache/pulsar/config/PropertyLoader.java (hunk 1 above)
Comment (gunnarmorling) [reply apache-pulsar/src/main/java/io/debezium/examples/apache/pulsar/config/PropertyLoader.java]: Adjusted the name; I think the other stuff isn't too important for a demo.

**Comment by gunnarmorling** [comment]
Pushed one commit to address @jpechane's remarks and merged. Thanks for reviewing to you!

**Pull request merged**